tests/e2e-tests/src/recursor/dnssec/scenarios.rs

```diff
@@ -0,0 +1,242 @@
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_servfail());
+    assert!(logs.contains("iteration count 501 is over 500"));
+
+    Ok(())
+}
+
+/// This tests the nsec3_soft_iteration_limit and nsec3_hard_iteration_limit configuration settings
+#[test]
+fn nsec3_custom_iteration_count() -> Result<()> {
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 11,
+            salt: None,
+            opt_out: false,
+        }),
+        Some(minijinja::render!(include_str!(
+            "custom_nsec3_iterations.toml.jinja"
+        ))),
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_nxdomain());
+    assert!(!output.flags.authenticated_data);
+    assert!(logs.contains("iteration count 11 is over 10"));
+
+    let (output, logs) = insecure_record_fixture(
+        FQDN("noexist.insecure.testing.")?,
+        SignSettings::default().nsec(Nsec::_3 {
+            iterations: 21,
+            salt: None,
+            opt_out: false,
+        }),
+        Some(minijinja::render!(include_str!(
+            "custom_nsec3_iterations.toml.jinja"
+        ))),
+    )?;
+
+    dbg!(&output);
+
+    assert!(output.status.is_servfail());
+    assert!(logs.contains("iteration count 21 is over 20"));
+
+    Ok(())
+}
+
+/// This test verifies the server will verify NSEC3 RRSIGs before rejecting high iteration counts.
+#[test]
+fn hard_nsec3_iteration_invalid_rrsig() -> Result<()> {
+    let network = Network::new()?;
+    let leaf_zone = FQDN::TEST_DOMAIN;
+    let mut leaf_ns = NameServer::new(&Implementation::test_peer(), leaf_zone.clone(), &network)?;
+
+    leaf_ns.add(Record::a(
+        leaf_zone.push_label("host"),
+        Ipv4Addr::new(1, 2, 3, 4),
+    ));
+
+    let Graph {
+        nameservers: _nameservers,
+        root,
+        trust_anchor,
+    } = Graph::build(
+        leaf_ns,
+        Sign::AndAmend {
+            settings: SignSettings::default().nsec(Nsec::_3 {
+                iterations: 501,
+                salt: None,
+                opt_out: false,
+            }),
+            mutate: &|zone, records| {
+                if zone == &leaf_zone {
+                    let mut did_remove = false;
+                    for record in records.iter_mut() {
+                        if let Record::RRSIG(rrsig) = record {
+                            if rrsig.type_covered == RecordType::NSEC3 {
+                                let mut rrsig = rrsig.clone();
+                                rrsig.signature = "bm90YXJlYWxycnNpZ3JlY29yZA==".to_string();
+                                *record = Record::RRSIG(rrsig);
+                                did_remove = true;
+                            }
+                        }
+                    }
+                    assert!(
+                        did_remove,
+                        "did not find an RRSIG covering DNSKEY generated using the KSK"
```

Comment:
nit: incorrect assertion message
```suggestion
                        "did not find an RRSIG covering an NSEC3 record"
```